### PR TITLE
[NEW] Star message, report and delete message events

### PR DESCRIPTION
**Pull request by tapiarafael**
# What? :boat:

Added definition and listeners for new events IPostMessageStarred, IPostMessageReported.
Added user info in the context of IPostMessageDeleted.

# Why? :thinking:

# Links :earth_americas:

[Task](https://app.clickup.com/t/22wqujz)
[Rocket.chat PR](https://github.com/RocketChat/Rocket.Chat/pull/25383)

# PS :eyes:

### CODE REVIEW

File: src/definition/messages/IMessageFollowContext.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    message: IMessage;
+    /**
+     * The user who follow the message
+     */
+    user: IUser;
+    /**
+     * If the message was followed or unfollowed
+     */
+    isUnfollow: boolean;
```
Comment (graywolf336) [inline src/definition/messages/IMessageFollowContext.ts:20]: Wouldn't this fall under the same principle as the other PR, no double negatives to make a positive?

File: src/definition/messages/IMessagePinContext.ts [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+     */
+    message: IMessage;
+    /**
+     * The user who pinned the message
+     */
+    user: IUser;
+    /**
+     * If the message was pinned or unpinned
+     */
+    isUnpinned: boolean;
```
Comment (graywolf336) [inline src/definition/messages/IMessagePinContext.ts:20]: Same here.

File: src/definition/messages/IMessageReactionContext.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ */
+export interface IMessageReactionContext {
+    /**
+     * The reaction itself
+     */
+    reaction: string;
+    /**
+     * If the reaction was removed or added from the message
+     */
+    isRemoved: boolean;
```
Comment (graywolf336) [inline src/definition/messages/IMessageReactionContext.ts:16]: And same here.

File: src/definition/messages/IPostMessageDeleted.ts [hunk 4]
```diff
@@ -8,20 +9,28 @@ export interface IPostMessageDeleted {
...
      * has been deleted.
      *
-     * @param message The message which was deleted
+     * @param message The deleted message
      * @param read An accessor to the environment
      * @param http An accessor to the outside world
      * @returns whether to run the executor function
      */
-    checkPostMessageDeleted?(message: IMessage, read: IRead, http: IHttp): Promise<boolean>;
+    checkPostMessageDeleted?(context: IMessageDeleteContext, read: IRead, http: IHttp): Promise<boolean>;
```
Comment (graywolf336) [inline src/definition/messages/IPostMessageDeleted.ts:17]: This is a major breaking change! Please do not do this!!!!

File: src/definition/metadata/AppMethod.ts [hunk 5]
```diff
@@ -24,8 +24,8 @@ export enum AppMethod {
     EXECUTEPOSTMESSAGESENT = 'executePostMessageSent',
     CHECKPREMESSAGEDELETEPREVENT = 'checkPreMessageDeletePrevent',
     EXECUTEPREMESSAGEDELETEPREVENT = 'executePreMessageDeletePrevent',
-    CHECKPOSTMESSAGEDELETED = 'checkPostMessageDeleted',
-    EXECUTEPOSTMESSAGEDELETED = 'executePostMessageDeleted',
+    CHECK_POST_MESSAGE_DELETED = 'checkPostMessageDeleted',
+    EXECUTE_POST_MESSAGE_DELETED = 'executePostMessageDeleted',
```
Comment (graywolf336) [inline src/definition/metadata/AppMethod.ts:28]: Another breaking change, please do not do this!!

File: src/server/managers/AppListenerManager.ts [hunk 6]
```diff
@@ -394,28 +410,30 @@ export class AppListenerManager {
...
         for (const appId of this.listeners.get(AppInterface.IPostMessageDeleted)) {
             const app = this.manager.getOneById(appId);
 
             let continueOn = true;
-            if (app.hasMethod(AppMethod.CHECKPOSTMESSAGEDELETED)) {
-                continueOn = await app.call(AppMethod.CHECKPOSTMESSAGEDELETED,
-                    cfMsg,
+            if (app.hasMethod(AppMethod.CHECK_POST_MESSAGE_DELETED)) {
+                continueOn = await app.call(AppMethod.CHECK_POST_MESSAGE_DELETED,
+                    message,
```
Comment (graywolf336) [inline src/server/managers/AppListenerManager.ts:423]: The definition above said this is context but you're still passing message. Continue passing the message here but you can add the context to the method's definition.

**Review by graywolf336 — DISMISSED** [review]
I've identified some major breaking changes that break compatibility with previous versions of the Apps.

File: src/definition/messages/IPostMessageDeleted.ts [hunk 7]
```diff
@@ -8,7 +9,7 @@ export interface IPostMessageDeleted {
      * this handler should actually be executed for after the message
      * has been deleted.
      *
-     * @param message The message which was deleted
+     * @param message The deleted message
      * @param read An accessor to the environment
      * @param http An accessor to the outside world
```
Comment (graywolf336) [inline src/definition/messages/IPostMessageDeleted.ts:14]: Any reason you didn't add the context here?

Comment (tapiarafael) [reply src/definition/messages/IPostMessageDeleted.ts:14]: I forgot. Adding right now!

File: src/server/managers/AppListenerManager.ts [hunk 8]
```diff
@@ -419,28 +434,31 @@ export class AppListenerManager {
...
+        const {message} = context;
 
         for (const appId of this.listeners.get(AppInterface.IPostMessageDeleted)) {
             const app = this.manager.getOneById(appId);
 
             let continueOn = true;
             if (app.hasMethod(AppMethod.CHECKPOSTMESSAGEDELETED)) {
                 continueOn = await app.call(AppMethod.CHECKPOSTMESSAGEDELETED,
-                    cfMsg,
+                    message,
```
Comment (d-gubert) [inline src/server/managers/AppListenerManager.ts:447]:
```suggestion
                    // `context` has more information about the event, but
                    // we had to keep this `message` here for compatibility
                    message,
```

**Review by d-gubert — APPROVED** [review]

**Pull request merged**